### PR TITLE
[WIP]: Support Bytes/Digest LHS for Modulo

### PR DESCRIPTION
**Pull request by chrisnevers**
# TODO
* Support Digests

### CODE REVIEW

File: docs/src/rsh/compute/index.md [hunk 1]
```diff
@@ -589,9 +589,10 @@ Numeric operations, like `{!rsh} +` and `{!rsh} >`, only operate on numbers.
 Since all numbers in Reach are integers, operations like `{!rsh} /` truncate their result.
 Boolean operations, like `{!rsh} &&`, only operate on booleans.
 `{!rsh} polyXor` operates on `{!rsh} UInt`s, `{!rsh} Bool`s, `{!rsh} Digest`s, and `{!rsh} Bytes` of the same length.
+`{!rsh} polyMod` operates on `{!rsh} UInt`s. The first argument may be a `{!rsh} Digest` or `{!rsh} Bytes` of at least 8 length.
```
Comment (jeapostrophe) [inline docs/src/rsh/compute/index.md:592]: You can make it work for anything length

File: hs/src/Reach/Connector/ETH_Solidity.hs [hunk 2]
```diff
@@ -965,6 +966,21 @@ solCom = \case
...
+    let (howMany, lastLen) = solBytesInfo $ bytesTypeLen $ argTypeOf x
+    let go elemIdx i = dv' <+> "=" <+> parens (dv' <+> "* 256") <+> "+" <+> "uint8" <> parens ("bytes1" <> parens (x' <> ei <> brackets (pretty i))) <> semi
+          where
+            ei = ".elem" <> pretty elemIdx
+    let (res :: [Doc]) = case lastLen < 8 of
+              -- Access all elements from last elem idx
+              False -> map (go $ howMany - 1) $ [lastLen - 8 .. lastLen - 1 :: Integer]
+              -- Access `lastLen` elements from `howMany` elem idx and `8 - lastLen` from `howMany - 1` elem idx
+              True -> map (go $ howMany - 2) [maxStringLen - lastLen .. maxStringLen - 1] <>
+                        map (go $ howMany - 1) [0 .. lastLen - 1]
```
Comment (chrisnevers) [inline hs/src/Reach/Connector/ETH_Solidity.hs:982]: Will make these output an actual `for` loop in solidity

File: examples/bytes-mod/index.mjs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+accAlice.setGasLimit(5000000);
+
+const ctcAlice = accAlice.contract(backend);
+
+const b = "ExampleeExampleeExampleeExampleeExam";
+
+await Promise.all([
+  backend.A(ctcAlice, {
+    b,
+    expected: stdlib.btoiLast8(b).mod(4)
```
Comment (jeapostrophe) [inline examples/bytes-mod/index.mjs:19]: This function should be in the docs?

File: hs/src/Reach/Connector/ALGO.hs [hunk 4]
```diff
@@ -1228,6 +1228,13 @@ cprim = \case
   DIGEST_EQ -> call "=="
   ADDRESS_EQ -> call "=="
   TOKEN_EQ -> call "=="
+  BTOI_LAST8 -> \case
+    [x] -> do
+      let len = fromIntegral $ bytesTypeLen $ argTypeOf x
+      ca x
+      output $ TExtract (len - 8) 8
```
Comment (jeapostrophe) [inline hs/src/Reach/Connector/ALGO.hs:1235]: Yea, you can make this work for any length

File: hs/src/Reach/Connector/ETH_Solidity.hs [hunk 5]
```diff
@@ -965,6 +966,21 @@ solCom = \case
...
+    x' <- solArg x
+    let (howMany, lastLen) = solBytesInfo $ bytesTypeLen $ argTypeOf x
+    let go elemIdx i = dv' <+> "=" <+> parens (dv' <+> "* 256") <+> "+" <+> "uint8" <> parens ("bytes1" <> parens (x' <> ei <> brackets (pretty i))) <> semi
+          where
+            ei = ".elem" <> pretty elemIdx
+    let (res :: [Doc]) = case lastLen < 8 of
+              -- Access all elements from last elem idx
+              False -> map (go $ howMany - 1) $ [lastLen - 8 .. lastLen - 1 :: Integer]
+              -- Access `lastLen` elements from `howMany` elem idx and `8 - lastLen` from `howMany - 1` elem idx
+              True -> map (go $ howMany - 2) [maxStringLen - lastLen .. maxStringLen - 1] <>
```
Comment (jeapostrophe) [inline hs/src/Reach/Connector/ETH_Solidity.hs:981]: I want more examples where you know that both of these paths get gone down

File: hs/src/Reach/Connector/ETH_Solidity.hs [hunk 6]
```diff
@@ -1230,22 +1246,29 @@ solHandlers (CHandlers hs) =
 divup :: Integer -> Integer -> Integer
 divup x y = ceiling $ (fromIntegral x :: Double) / (fromIntegral y)
 
-solBytesSplit :: Integer -> (Integer -> Integer -> a) -> [a]
-solBytesSplit sz f = map go [0 .. lastOne]
+maxStringLen :: Integer
+maxStringLen = 32
```
Comment (jeapostrophe) [inline hs/src/Reach/Connector/ETH_Solidity.hs:1250]: This is a deceptive name... it is like the size of the chunks... so `byteChunkSize`?

File: hs/src/Reach/Eval/Core.hs [hunk 7]
```diff
@@ -3463,6 +3470,17 @@ evalPrim p sargs =
         (l, r) -> expect_ $ Err_xor_Types l r
       where
         prim = flip evalPrimOp sargs
+    SLPrim_mod -> do
+      (x, y) <- two_args
+      (x_ty, _) <- compileTypeOf x
+      (y_ty, _) <- compileTypeOf y
+      case (x_ty, y_ty) of
+        (T_UInt, T_UInt) -> evalPrimOp MOD sargs
+        (T_Bytes sz, T_UInt)
```
Comment (jeapostrophe) [inline hs/src/Reach/Eval/Core.hs:3479]: You say in the docs it works on Digests, but there's no test for that and you don't allow it here

**Review by jeapostrophe — CHANGES_REQUESTED** [review]

File: hs/src/Reach/Eval/Core.hs (hunk 7 above)
Comment (chrisnevers) [reply hs/src/Reach/Eval/Core.hs:3479]: See my initial PR comment 😉

Comment (jeapostrophe) [reply hs/src/Reach/Eval/Core.hs:3479]: Sorry. I don't read :)

**Review by jeapostrophe — CHANGES_REQUESTED** [review]
changelog

File: hs/src/Reach/Connector/ALGO.hs [hunk 8]
```diff
@@ -1321,6 +1321,16 @@ cprim = \case
   DIGEST_EQ -> call "=="
   ADDRESS_EQ -> call "=="
   TOKEN_EQ -> call "=="
+  BTOI_LAST8 isDigest -> \case
+    [x] -> do
+      let bl = fromIntegral $ case isDigest of
+                False -> bytesTypeLen $ argTypeOf x
+                True  -> typeSizeOf T_Digest
```
Comment (jeapostrophe) [inline hs/src/Reach/Connector/ALGO.hs:1328]: These can both just be `typeSizeOf $ argTypeOf x`

**Review by jeapostrophe — CHANGES_REQUESTED** [review]

**Review by jeapostrophe — APPROVED** [review]

**Pull request merged**